### PR TITLE
chore: Migrated ssm, sts, fis, and sqs to v2

### PR DESCRIPTION
**Pull request by edibble21**
Fixes #N/A 

**Description**
Migrated [ssm](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ssm), [sts](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/sts), [fis](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/fis), and [sqs](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/sqs) to v2
**How was this change tested?**
make test
e2e tests
**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by edibble21 — COMMENTED** [review]
/Karpenter snapshot

**Review by edibble21 — COMMENTED** [review]
/Karpenter snapshot

File: pkg/aws/sdk.go [hunk 1]
```diff
@@ -29,4 +33,30 @@ type IAMAPI interface {
...
+	DeleteMessage(context.Context, *sqs.DeleteMessageInput, ...func(*sqs.Options)) (*sqs.DeleteMessageOutput, error)
+	SendMessage(context.Context, *sqs.SendMessageInput, ...func(*sqs.Options)) (*sqs.SendMessageOutput, error)
+}
+
+type STSAPI interface {
+	// STS Methods
+	GetCallerIdentity(context.Context, *sts.GetCallerIdentityInput, ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error)
+}
+
+type FISAPI interface {
```
Comment (engedaam) [inline pkg/aws/sdk.go:57]: We should move this `test` package

File: pkg/controllers/controllers.go [hunk 2]
```diff
@@ -69,8 +71,9 @@ func NewControllers(ctx context.Context, mgr manager.Manager, sess *session.Sess
 		status.NewController[*v1.EC2NodeClass](kubeClient, mgr.GetEventRecorderFor("karpenter")),
 	}
 	if options.FromContext(ctx).InterruptionQueue != "" {
-		sqsapi := servicesqs.New(sess)
-		out := lo.Must(sqsapi.GetQueueUrlWithContext(ctx, &servicesqs.GetQueueUrlInput{QueueName: lo.ToPtr(options.FromContext(ctx).InterruptionQueue)}))
+		cfg := lo.Must(config.LoadDefaultConfig(context.Background(), config.WithRetryMaxAttempts(3)))
```
Comment (engedaam) [inline pkg/controllers/controllers.go:74]:
```suggestion
		cfg := lo.Must(config.LoadDefaultConfig(ctx, config.WithRetryMaxAttempts(3)))
```

File: test/pkg/environment/aws/environment.go [hunk 3]
```diff
@@ -118,8 +123,9 @@ func NewEnvironment(t *testing.T) *Environment {
 	}
 	// Initialize the provider only if the INTERRUPTION_QUEUE environment variable is defined
 	if v, ok := os.LookupEnv("INTERRUPTION_QUEUE"); ok {
-		sqsapi := servicesqs.New(session)
-		out := lo.Must(sqsapi.GetQueueUrlWithContext(env.Context, &servicesqs.GetQueueUrlInput{QueueName: aws.String(v)}))
+		cfg := lo.Must(config.LoadDefaultConfig(context.Background(), config.WithRetryMaxAttempts(3)))
```
Comment (engedaam) [inline test/pkg/environment/aws/environment.go:126]: Why do we need to re-define the `cfg`?
```suggestion
		cfg := lo.Must(config.LoadDefaultConfig(ctx, config.WithRetryMaxAttempts(3)))
```

File: test/pkg/environment/aws/environment.go [hunk 4]
```diff
@@ -15,24 +15,25 @@ limitations under the License.
...
 	"testing"
 
 	coretest "sigs.k8s.io/karpenter/pkg/test"
 
+	"github.com/aws/aws-sdk-go-v2/service/fis"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	servicesqs "github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/aws/aws-sdk-go-v2/service/ssm"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
 	"github.com/aws/aws-sdk-go/aws"
```
Comment (engedaam) [inline test/pkg/environment/aws/environment.go:30]: Shouldn't we drop most of these?

File: test/pkg/environment/aws/environment.go [hunk 5]
```diff
@@ -60,11 +64,11 @@ type Environment struct {
...
+	STSAPI        sdk.STSAPI
 	EC2API        *ec2.EC2
-	SSMAPI        *ssm.SSM
-	IAMAPI        *iam.IAM
-	FISAPI        *fis.FIS
+	SSMAPI        sdk.SSMAPI
+	IAMAPI        sdk.IAMAPI
+	FISAPI        sdk.FISAPI
 	EKSAPI        *eks.EKS
 	TimeStreamAPI timestreamwriteiface.TimestreamWriteAPI
```
Comment (engedaam) [inline test/pkg/environment/aws/environment.go:73]: Can we update timestream as well?

File: test/pkg/environment/aws/environment.go [hunk 6]
```diff
@@ -118,8 +123,9 @@ func NewEnvironment(t *testing.T) *Environment {
 	}
 	// Initialize the provider only if the INTERRUPTION_QUEUE environment variable is defined
 	if v, ok := os.LookupEnv("INTERRUPTION_QUEUE"); ok {
-		sqsapi := servicesqs.New(session)
-		out := lo.Must(sqsapi.GetQueueUrlWithContext(env.Context, &servicesqs.GetQueueUrlInput{QueueName: aws.String(v)}))
+		cfg := lo.Must(config.LoadDefaultConfig(context.Background(), config.WithRetryMaxAttempts(3)))
+		sqsapi := servicesqs.NewFromConfig(cfg)
+		out := lo.Must(sqsapi.GetQueueUrl(env.Context, &servicesqs.GetQueueUrlInput{QueueName: aws.String(v)}))
```
Comment (engedaam) [inline test/pkg/environment/aws/environment.go:128]:
```suggestion
		out := lo.Must(sqsapi.GetQueueUrl(env.Context, &servicesqs.GetQueueUrlInput{QueueName: aws.ToString(v)}))
```

File: test/pkg/environment/aws/expectations.go [hunk 7]
```diff
@@ -355,7 +358,7 @@ func (env *Environment) K8sMinorVersion() int {
 func (env *Environment) GetAMIBySSMPath(ssmPath string) string {
 	GinkgoHelper()
 
-	parameter, err := env.SSMAPI.GetParameter(&ssm.GetParameterInput{
+	parameter, err := env.SSMAPI.GetParameter(context.Background(), &ssm.GetParameterInput{
```
Comment (engedaam) [inline test/pkg/environment/aws/expectations.go:361]:
```suggestion
	parameter, err := env.SSMAPI.GetParameter(env.Context, &ssm.GetParameterInput{
```

File: test/pkg/environment/aws/expectations.go [hunk 8]
```diff
@@ -389,30 +392,30 @@ func (env *Environment) ExpectSpotInterruptionRole() *iam.Role {
...
+		Tags: []iamtypes.Tag{
 			{
 				Key:   aws.String(coretest.DiscoveryLabel),
 				Value: aws.String(env.ClusterName),
 			},
 		},
 	}
 	By("adding the karpenter role to new instance profile")
-	_, err := env.IAMAPI.CreateInstanceProfile(createInstanceProfile)
+	_, err := env.IAMAPI.CreateInstanceProfile(context.Background(), createInstanceProfile)
```
Comment (engedaam) [inline test/pkg/environment/aws/expectations.go:412]:
```suggestion
	_, err := env.IAMAPI.CreateInstanceProfile(env.Context, createInstanceProfile)
```

File: test/pkg/environment/aws/expectations.go [hunk 9]
```diff
@@ -389,30 +392,30 @@ func (env *Environment) ExpectSpotInterruptionRole() *iam.Role {
...
 	Expect(awserrors.IgnoreAlreadyExists(err)).ToNot(HaveOccurred())
 	addInstanceProfile := &iam.AddRoleToInstanceProfileInput{
 		InstanceProfileName: aws.String(instanceProfileName),
 		RoleName:            aws.String(roleName),
 	}
-	_, err = env.IAMAPI.AddRoleToInstanceProfile(addInstanceProfile)
+	_, err = env.IAMAPI.AddRoleToInstanceProfile(context.Background(), addInstanceProfile)
```
Comment (engedaam) [inline test/pkg/environment/aws/expectations.go:418]: Let's make sure and update all the E2E tests to use env.Context

**Review by engedaam — COMMENTED** [review]
Few nits

File: test/pkg/environment/aws/environment.go (hunk 6 above)
Comment (edibble21) [reply test/pkg/environment/aws/environment.go:128]: aws.String takes a string and makes it a *string while aws.ToString takes a *string and makes it a string and QueueName is expecting *string so aws.String would be correct. I did migrate it to v2 though.

File: test/pkg/environment/aws/environment.go (hunk 4 above)
Comment (edibble21) [reply test/pkg/environment/aws/environment.go:30]: They are used later in the file to define the new clients

File: test/pkg/environment/aws/environment.go [hunk 10]
```diff
@@ -60,11 +65,11 @@ type Environment struct {
 	*common.Environment
 	Region string
 
-	STSAPI        *sts.STS
+	STSAPI        sdk.STSAPI
```
Comment (engedaam) [inline test/pkg/environment/aws/environment.go:68]: nit. You should be able to us the EC2 client option for all of these clients

File: test/suites/integration/tags_test.go [hunk 11]
```diff
@@ -106,10 +108,10 @@ var _ = Describe("Tags", func() {
 			env.ExpectCreated(nodeClass)
 			env.EventuallyExpectInstanceProfileExists(env.GetInstanceProfileName(nodeClass))
 
-			_, err := env.IAMAPI.UntagInstanceProfile(&iam.UntagInstanceProfileInput{
+			_, err := env.IAMAPI.UntagInstanceProfile(context.Background(), &iam.UntagInstanceProfileInput{
```
Comment (engedaam) [inline test/suites/integration/tags_test.go:111]:
```suggestion
			_, err := env.IAMAPI.UntagInstanceProfile(env.Context, &iam.UntagInstanceProfileInput{
```

File: test/suites/integration/tags_test.go [hunk 12]
```diff
@@ -118,11 +120,11 @@ var _ = Describe("Tags", func() {
 			env.EventuallyExpectKarpenterRestarted()
 
 			Eventually(func(g Gomega) {
-				out, err := env.IAMAPI.GetInstanceProfile(&iam.GetInstanceProfileInput{
+				out, err := env.IAMAPI.GetInstanceProfile(context.Background(), &iam.GetInstanceProfileInput{
```
Comment (engedaam) [inline test/suites/integration/tags_test.go:123]:
```suggestion
				out, err := env.IAMAPI.GetInstanceProfile(env.Context, &iam.GetInstanceProfileInput{
```

**Review by engedaam — APPROVED** [review]
LGTM 🚀

**Pull request merged**